### PR TITLE
:recycle:(seg metrics)! rework and add PAvPU

### PR DESCRIPTION
**Pull request by alafage**
This PR introduces a wrapper `SegmentationMetric` for `torchmetrics.Metric` objects. Its purpose is to provide utilities for simple metric computation over images:

- [x] Masking
- [x] Subsampling
- [ ] Add PAvPU
  - [x] Implement PAvPU metric
  - [x] Integrate PAvPU into the routine
  - [ ] Compute the uncertainty threshold dynamically (see https://arxiv.org/pdf/1811.12709) (no clean solution yet)
- [ ] Patch-based computation (no solution yet)

### CODE REVIEW

File: src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        """Update the metric with new predictions and targets.
+
+        Args:
+            preds: Tensor of shape (N, C, H, W) containing the predicted probabilities or logits.
+            target: Tensor of shape (N, H, W) or (N, 1, H, W) containing the ground truth labels.
+            ignore_mask: Optional tensor of shape (N, H, W) or (N, 1, H, W) indicating which pixels to ignore.
+        """
+        preds, target, valid = self._validate_shapes(preds, target, ignore_mask)
+
+        if not torch.all((preds >= 0) & (preds <= 1)):
```
Comment (o-laurent) [inline src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py:77]: Maybe log that we apply the softmax since it s non-std in the lib?

File: src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+from torchmetrics import Metric
+
+
+class PAvPU(Metric):
```
Comment (o-laurent) [inline src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py:8]: Do we have to set the défaut values for higher_is_better etc?

File: src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        if not torch.all((preds >= 0) & (preds <= 1)):
+            preds = F.softmax(preds, dim=1)
+
+        hard_preds = preds.argmax(dim=1)
+        acc_map = (hard_preds == target).float()
+        unc_map = 1 - preds.max(dim=1).values
+
+        ps = self.patch_size
+        valid_f = valid.float().unsqueeze(1)
+        pooled_valid = F.avg_pool2d(
```
Comment (o-laurent) [inline src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py:91]: Maybe add a comment (patchify)

File: src/torch_uncertainty/metrics/segmentation/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    def update(self, preds: Tensor, target: Tensor, ignore_mask: Tensor | None = None):
+        """Standard update method for segmentation metrics.
+
+        Args:
+            preds: logits or probabilities of shape (N, C, H, W).
+            target: ground truth labels of shape (N, H, W) or (N, 1, H, W).
+            ignore_mask: mask to ignore certain pixels in the computation
+                of size (N, H, W) or (N, 1, H, W). Defaults to ``None``.
+        """
+        if target.ndim == 4:
```
Comment (o-laurent) [inline src/torch_uncertainty/metrics/segmentation/wrapper.py:27]: Can we not use this logic in PavPu?

File: src/torch_uncertainty/routines/segmentation.py [hunk 5]
```diff
@@ -90,6 +95,9 @@ def __init__(
                 is only used if :attr:`log_plots` is set to ``True``. Defaults to ``3``.
             num_bins_calibration_error: Number of bins to compute calibration error metrics.
                 Defaults to ``15``.
+            patch_size: The size of the patches to compute the PAvPU metric. Defaults to ``16``.
```
Comment (o-laurent) [inline src/torch_uncertainty/routines/segmentation.py]: Maybe precise pavpu_patch_size and the same for the two arguments below?

**Review by o-laurent — APPROVED** [review]

File: src/torch_uncertainty/routines/segmentation.py [hunk 6]
```diff
@@ -369,7 +388,7 @@ def test_step(
             labels[id_mask] = 0  # ID examples
             labels[ood_mask] = 1  # OOD examples
 
-            self.test_ood_metrics.update(ood_scores, labels)
+            self.test_ood_metrics.update(ood_scores, labels, ignore_mask=ignore_mask)
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply src/torch_uncertainty/routines/segmentation.py]: @alafage have you tried this?

File: src/torch_uncertainty/routines/segmentation.py [hunk 7]
```diff
@@ -389,6 +408,7 @@ def on_test_epoch_end(self) -> None:
         """Compute, log, and plot the values of the collected metrics in `test_step`."""
         result_dict = self.test_seg_metrics.compute()
         result_dict |= self.test_sbsmpl_seg_metrics.compute()
+        result_dict |= self.test_patch_seg_metrics.compute()
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply src/torch_uncertainty/routines/segmentation.py:418]: This comment seems legit

File: src/torch_uncertainty/routines/segmentation.py [hunk 8]
```diff
@@ -415,16 +435,16 @@ def on_test_epoch_end(self) -> None:
     def _plot_results(self):
         """Plot uncertainty quantification metrics and segmentation figures."""
         self.logger.experiment.add_figure(
-            "Calibration/Reliability diagram",
-            self.test_sbsmpl_seg_metrics["cal/ECE"].plot()[0],
+            "Calibration/Reliabity diagram",
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply src/torch_uncertainty/routines/segmentation.py]: @copilot write a suggestion that we can apply directly

**Review by o-laurent — COMMENTED** [review]
Thanks for your work @alafage. Copilot's comments seem legit.

File: src/torch_uncertainty/metrics/segmentation/wrapper.py (hunk 4 above)
Comment (alafage) [reply src/torch_uncertainty/metrics/segmentation/wrapper.py:27]: The same logic is done by: `preds, target, valid = self._validate_shapes(preds, target, ignore_mask)` within the update.

File: src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py (hunk 2 above)
Comment (alafage) [reply src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py:8]: You're right, I added the missing default values

File: src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py (hunk 1 above)
Comment (alafage) [reply src/torch_uncertainty/metrics/segmentation/patch_accuracy_vs_uncertainty.py:77]: I added a warning.

File: src/torch_uncertainty/routines/segmentation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alafage) [reply src/torch_uncertainty/routines/segmentation.py]: This comment was correct, I fixed it
